scipy/linalg/_sketches.py

```diff
@@ -35,42 +37,38 @@ def cwt_matrix(n_rows, n_columns, seed=None):
 
     Returns
     -------
-    S : (n_rows, n_columns) array_like
+    S : (n_rows, n_columns) csc_matrix
+        The returned matrix has ``n_columns`` nonzero entries.
 
     Notes
     -----
     Given a matrix A, with probability at least 9/10,
-    .. math:: ||SA|| == (1 \pm \epsilon)||A||
-    Where epsilon is related to the size of S
+    .. math:: \|SA\| = (1 \pm \epsilon)\|A\|
+    Where the error epsilon is related to the size of S
     """
-    S = np.zeros((n_rows, n_columns))
-    nz_positions = np.random.randint(0, n_rows, n_columns)
     rng = check_random_state(seed)
-    values = rng.choice([1, -1], n_columns)
-    for i in range(n_columns):
-        S[nz_positions[i]][i] = values[i]
-
-    return S
+    rows = rng.randint(0,n_rows,n_columns)
+    cols = np.arange(n_columns)
+    signs = rng.choice([1, -1], n_columns)
+    S = coo_matrix((signs,(rows,cols)),shape=(n_rows,n_columns))
+    return S.tocsc()
```

Comment:
You can use the `(data, (rows, cols))` version of the `csc_matrix` constructor here and avoid the `.tocsc()` call.
http://scipy.github.io/devdocs/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix

Reply:
Thanks for letting me know about this! I didn't know `csc_matrix` had such a constructor.

Nevertheless, and perhaps surprisingly, the current method of constructing a `coo_matrix` and then converting to a `csc_matrix` is anywhere from 29 to 144 percent faster at creating the sketching matrix than the proposed construction just using `csc_matrix`. (See the benchmark code below, and perhaps someone can run themselves to check that this isn't specific to my machine; I'm running `scipy.__version__=1.2.1` and `numpy.__version__=1.15.4`.)

Does anyone know a good reason for this other than the fact that `coo_matrix` is meant to be fast at construction, as it says in the docs? Should this be a separate issue in it's own right?

-------
```py
import numpy as np
from scipy import sparse

def sketch_coo(k=10,n=100):
    '''sketch_coo
    Implemenents Clarkson Woodruff sketch matrix
    as a scipy.sparse.csc sparse matrix, first constructing
    the result as a scipy.sparse.coo matrix.
    '''
    i = np.random.randint(0,k,n)
    j = np.arange(n)
    data = np.random.choice([1, -1], n)
    return sparse.coo_matrix((data,(i,j)),shape=(k,n)).tocsc()

def sketch_csc(k=10,n=100):
    '''sketch_csc
    Implemenents Clarkson Woodruff sketch matrix
    as a scipy.sparse.csr sparse matrix directly.
    '''
    i = np.random.randint(0,k,n)
    j = np.arange(n)
    data = np.random.choice([1, -1], n)
    return sparse.csr_matrix((data,(i,j)),shape=(k,n))

print('\nk=10, n=1500')
%timeit sketch_coo(k=10, n=1500)
%timeit sketch_csc(k=10, n=1500)
print('\nk=100, n=1500')
%timeit sketch_coo(k=100, n=1500)
%timeit sketch_csc(k=100, n=1500)
print('\nk=250, n=1500')
%timeit sketch_coo(k=100, n=1500)
%timeit sketch_csc(k=100, n=1500)
print('\nk=1000, n=1500')
%timeit sketch_coo(k=1000, n=1500)
%timeit sketch_csc(k=1000, n=1500)
print('\nk=250, n=15000')
%timeit sketch_coo(k=100, n=1500)
%timeit sketch_csc(k=100, n=1500)
```
gives output
```
k=10, n=1500
199 µs ± 10.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
291 µs ± 26.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

k=100, n=1500
210 µs ± 576 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
271 µs ± 5.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

k=250, n=1500
210 µs ± 8.08 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
318 µs ± 5.38 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

k=1000, n=1500
232 µs ± 8.87 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
321 µs ± 15.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

k=250, n=15000
260 µs ± 17.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
376 µs ± 55.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
Interesting, it looks like the "direct" version is incurring about ~100us of fixed overhead cost due to the circuitous way that code path is implemented internally. That can certainly be fixed, but there's actually a better way for your specific case.

Remember that CSC format stores its data and row-indices in the exact same format as COO, and only the column-indices differ. Due to the way you construct the matrix in this code, the CSR `indptr` array is trivial: `np.arange(n+1)` (that is, just one more element than the `j` column index array for COO).

With this change applied to `sketch_csc`, I get:
```
In [41]: %timeit sketch_coo(k=1000, n=1500)
10000 loops, best of 3: 191 µs per loop

In [42]: %timeit sketch_csc(k=1000, n=1500)
10000 loops, best of 3: 74.5 µs per loop
```

Reply:
Great point @perimosocordiae. I verified the speedup on my machine and updated to this construction in the most recent commit. Thanks!